src/dbnode/storage/namespace/schema.go

```diff
@@ -0,0 +1,226 @@
+	errSchemaNotFound       = errors.New("schema is not found")
+	errSchemaRegistryEmpty  = errors.New("schema registry is empty")
+	errInvalidSchemaOptions = errors.New("invalid schema options")
+)
+
+type schemaDescr struct {
+	deployId string
+	md       *desc.MessageDescriptor
+}
+
+func (s *schemaDescr) DeployId() string {
+	return s.deployId
+}
+
+func (s *schemaDescr) Equal(o SchemaDescr) bool {
+	if s == nil && o == nil {
+		return true
+	}
+	if s != nil && o == nil || s == nil && o != nil {
+		return false
+	}
+	if _, ok := o.(*schemaDescr); !ok {
+		return false
+	}
+	return s.DeployId() == o.DeployId()
+}
+
+func (s *schemaDescr) Get() *desc.MessageDescriptor {
+	return s.md
+}
+
+func (s *schemaDescr) String() string {
+	if s.md == nil {
+		return ""
+	}
+	return s.md.String()
+}
+
+type schemaRegistry struct {
+	options  *nsproto.SchemaOptions
+	latestId string
+	// a map of schema version to schema descriptor.
+	versions map[string]*schemaDescr
+}
+
+func (sr *schemaRegistry) Equal(o SchemaRegistry) bool {
+	var osr *schemaRegistry
+	var ok bool
+	if sr == nil && o == nil {
+		return true
+	}
+	if sr != nil && o == nil || sr == nil && o != nil {
+		return false
+	}
+
+	if osr, ok = o.(*schemaRegistry); !ok {
+		return false
+	}
+	// compare latest version
+	if sr.latestId != osr.latestId {
+		return false
+	}
+
+	// compare version map
+	if len(sr.versions) != len(osr.versions) {
+		return false
+	}
+	for v, sd := range sr.versions {
+		osd, ok := osr.versions[v]
+		if !ok {
+			return false
+		}
+		if !sd.Equal(osd) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (sr *schemaRegistry) Get(id string) (SchemaDescr, error) {
+	sd, ok := sr.versions[id]
+	if !ok {
+		return nil, errSchemaNotFound
+	}
+	return sd, nil
+}
+
+func (sr *schemaRegistry) GetLatest() (SchemaDescr, error) {
+	return sr.Get(sr.latestId)
+}
+
+// toSchemaOptions returns the corresponding SchemaOptions proto for the provided SchemaRegistry
+func toSchemaOptions(sr SchemaRegistry) *nsproto.SchemaOptions {
+	if sr == nil {
+		return nil
+	}
+	_, ok := sr.(*schemaRegistry)
+	if !ok {
+		return nil
+	}
+	return sr.(*schemaRegistry).options
+}
+
+func emptySchemaRegistry() SchemaRegistry {
+	return &schemaRegistry{options: nil, versions: make(map[string]*schemaDescr)}
+}
+
+// LoadSchemaRegistry loads schema registry from SchemaOptions proto.
+func LoadSchemaRegistry(options *nsproto.SchemaOptions) (SchemaRegistry, error) {
+	sr := &schemaRegistry{options: options, versions: make(map[string]*schemaDescr)}
+	if options == nil ||
+		options.GetHistory() == nil ||
+		len(options.GetHistory().GetVersions()) == 0 {
+		return sr, nil
+	}
+
+	msgName := options.GetDefaultMessageName()
+	if len(msgName) == 0 {
+		return nil, xerrors.Wrap(errInvalidSchemaOptions, "default message name is not specified")
+	}
+
+	var prevId string
+	for _, fdbSet := range options.GetHistory().GetVersions() {
+		if len(prevId) > 0 && fdbSet.PrevId != prevId {
```

Comment:
presumably a zero length ID is an error? Or is that allowed?

Reply:
the first schema do not have prevId.